lib/syscall_shim: Ignore only unnumbered legacy syscalls

razvanvirtan opened the pull request:
### Prerequisite checklist
 - [X] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [X] Tested your changes against relevant architectures and platforms;
 - [X] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [X] Updated relevant documentation.


### Base target
 - Architecture(s): [`x86_64`, `arm64`]
 - Platform(s): [e.g. `kvm`]

### Description of changes

Currently, all the unnumbered system calls are ignored during the build process.
This shouldn't be the case for architecture specific syscalls. Those should
generate the failure of the build process if their numbere is not defined
for the corresponding architecture (see arch_prctl).
This commit introduces a separate header file that allows us to only ignore
unnumbered legacy syscalls from the architecture specific ones.

Currently, all the syscalls that caused problems on ARM64 are legacy syscalls.
However, when architecture specific syscalls will be introduced, those should be placed
between corresponding guards, as in [1].

[1] https://github.com/unikraft/unikraft/pull/336

Signed-off-by: Răzvan Vîrtan <virtanrazvan@gmail.com>

razvanvirtan committed:
lib/syscall_shim: Ignore only unnumbered legacy syscalls

Currently, all the unnumbered system calls are ignored
during the build process. This shouldn't be the case
for architecture specific syscalls. Those should generate
the failure of the build process if their number is not defined
for the corresponding architecture (see arch_prctl).
This commit introduces a separate header file that allows
us to only ignore unnumbered legacy syscalls.

Signed-off-by: Răzvan Vîrtan <virtanrazvan@gmail.com>

razvand commented on lib/syscall_shim/gen_provided.awk:
```diff
@@ -7,7 +7,11 @@ BEGIN {
 /[a-zA-Z0-9]+-[0-9]+/{
 	printf "\n#ifndef SYS_%s\n", $1;
-	printf "#warning Ignoring system call '%s': No system call number available\n", $1
+	printf "#ifdef LEGACY_%s\n", $1;
```
Could you add a short explanation of what this does? If the `LEGACY_...` macro is defined then that means there is a legacy system call that is ignored?

So a system call is either:
* required and defined: all OK
* not required and not-defined (i.e. legacy), warning shown
* required and not-defined, error shown

Is this correct?

razvanvirtan replied:
Yes, exactly. I will add the explanation.

razvand commented on lib/syscall_shim/include/uk/legacy_syscall.h:
```diff
@@ -0,0 +1,27 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
```
How did you create this list of system calls? Do they pertain to `aarch64`?

razvanvirtan replied:
Actually, the system calls added here are those that are not available on `aarch64`.

I choose the legacy syscalls based on the following criteria:
1. there is a newer syscall that includes their functionality
2. they are not architecture specific (i.e. they are not available only on x86, but on a wider set of architectures) [1]

For each of the system calls added here as legacy, I checked the man pages for alternatives:
 - all the syscalls provided by `lib/vfscore` that were included here have a newer alternative that can provide their functionality (e.g. `open()` -> `openat()`, `access()` -> `faccessat()` etc)
 - `fork()`, `getpgrp()` -> `clone()`, `getpgid()`
 - `time()`, `alarm()`, `pause()` -> `gettimeofday()`, `timer_settime()`, `sigsuspend()`
 - for `pause()`, see also [2]
 
 [1] https://marcin.juszkiewicz.com.pl/download/tables/syscalls.html
 [2] https://stackoverflow.com/questions/6328055/whats-the-problem-of-pause-at-all

razvand commented on lib/syscall_shim/include/uk/legacy_syscall.h:
```diff
@@ -0,0 +1,27 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
```
Shouldn't this be `BSD-3-Clause`?

razvanvirtan replied:
Actually I am not an expert into this, I see that we have BSD-2 on some files (e.g. [1])  and BSD-3 on other files  (e.g. [2])  . For now I will switch to BSD-3 here for consistency in the same directory (and also add the missing copyright in the header).

[1] https://github.com/unikraft/unikraft/blob/staging/arch/x86/x86_64/include/uk/asm/limits.h
[2] https://github.com/unikraft/unikraft/blob/staging/lib/syscall_shim/include/uk/syscall.h

razvand reviewed (commented):
Thanks, @razvanvirtan . See my comments.

Also, there is a typo in the commit message: `numbere` instead of `number`.

razvanvirtan commented:
@razvand I have applied your comments in this update.
- switched from BSD-2 to BSD-3 in the legacy header (see also the conversation)
- added some comments to `gen_provided.awk`
- corrected the typo in the commit description

razvand reviewed (approved):
Thanks, @razvanvirtan , this looks OK with me.

Let's get @skuenzer 's input and we can then approve and merge this PR.

skuenzer commented on lib/syscall_shim/gen_provided.awk:
```diff
@@ -6,8 +6,17 @@ BEGIN {
+	# if the LEGACY symbol is not defined for this syscall, it means that the
+	# syscall is required and we can't continue without a definition
+	printf "#else\n";
+	printf "#error System call %s not implemented\n", $1
```
Can we change this error message to be slightly more verbose, like
```awk
printf "#error Failed to map system call '%s': No system call number available\n", $1
```

skuenzer commented on lib/syscall_shim/include/uk/legacy_syscall.h:
```diff
@@ -0,0 +1,65 @@
+#define LEGACY_stat
+#define LEGACY_getdents
+#define LEGACY_time
+#define LEGACY_futimesat
```
Can you add a comment which syscall should be used instead (at least for the non-obvious), like:
```c
#define LEGACY_SYS_futimesat /* modern: utimensat */
```

skuenzer commented on lib/syscall_shim/include/uk/legacy_syscall.h:
```diff
@@ -0,0 +1,65 @@
+ */
+#ifndef __UK_LEGACY_SYSCALL_H__
+#define __UK_LEGACY_SYSCALL_H__
+
```
I'd prefer if we namespace the defines with `LEGACY_SYS_*`, they would be then closer connected to `SYS_*` defines.

skuenzer replied:
Can you add also a comment with a short description what this list of syscalls is about? One or two sentences explaining what a legacy syscall is.

skuenzer replied:
Can you also maybe add as comment behind a legacy definition that tells what the modern version is. For the less obvious cases this could be helpful, e.g.:
```c
#define LEGACY_SYS_futimesat /* modern: utimensat */
```

skuenzer reviewed (changes requested):
Hey Razvan,
this PR is cleverly solved :wink:. Since it prints always a warning if a mapping could not be found for a legacy system call, I think it is okay for this release. However, we should probably  change the behavior in the future:
My idea was that legacy system call wrappers would be registered to a separate list with the `Makefile.uk`. Instead of going to `UK_PROVIDED_SYSCALLS`, they would go to `UK_LEGACY_SYSCALLS`. This way we can make sure the syscall_shim automatically confirms with the intention of an implementation. If a wrapper is not ported yet, such a call would still stay in `UK_PROVIDED_SYSCALLS` and would mean that the build system should emit a warning or error.

Can you integrate the cosmetic change requests that I commented inline? Afterwards, we can take it upstream.

Thanks,

Simon

kubanrob commented:
At least on aarch64, this generates a large amount of warnings in every single file that includes the generated header. This buries any other warning.

It would be nice to have a configuration option to not generate the warnings.

skuenzer commented:
> 



> At least on aarch64, this generates a large amount of warnings in every single file that includes the generated header. This buries any other warning.
> 
> It would be nice to have a configuration option to not generate the warnings.

Good point actually. I think we should properly fix this after this release, given the time we have left. In general, a warning should then only be printed for implementations that do not have a modern syscall yet. I expect another commit to fix the regsitrations of some of the system calls and to provide the proper compatibility (legacy) wrappers.
However, a small fix with a config option that is on as default (so default the warning is emitted) - as @kubanrob suggested - could be a quick work around until we have it properly implemented. What do you think @razvanvirtan ?

razvanvirtan commented:
@skuenzer @kubanrob Done, I have applied the comments and added a new config option to the `syscall-shim` submenu to suppress the warnings. It works as expected on my side.

razvanvirtan commented:
@skuenzer @kubanrob Done, I have applied the comments and added a new config option to the `syscall-shim` submenu to suppress the warnings. It works as expected on my side.

kubanrob commented:
> added a new config option to the `syscall-shim` submenu to suppress the warnings.

The config option works great for me, thank you for adding it. I experienced no problems using this PR with aarch64 either.

skuenzer commented on lib/syscall_shim/Config.uk:
```diff
@@ -39,4 +39,11 @@ if LIBSYSCALL_SHIM
 		select LIBUKDEBUG_PRINTD
+
+	config LIBSYSCALL_SHIM_LEGACY_VERBOSE
+		bool "Warn for unimplemented legacy syscalls"
```
Better to call it unmapped, removed, suppressed or similar instead of unimplemented - this might be confusing.

skuenzer reviewed (changes requested):
Looks really good, I would just adopt the wording of the config option to avoid confusion.

Thanks,
Simon

razvanvirtan commented:
@skuenzer Done, I changed the config option text.

skuenzer commented:
Perfect :+1:. This sounds much better now. Can you fix the checkpatch warning? You probably just need to shorten the commit description lines to 75 characters.

razvanvirtan commented:
@skuenzer Sorry, I missed that, should be fixed now.

razvand reviewed (approved):
All OK. Thanks, @razvanvirtan 

Reviewed-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>

skuenzer reviewed (approved):
Approved-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
